pkg/machine/e2e/cp_test.go

```diff
@@ -1,18 +1,210 @@
+		)
+
+		sourceDir, err := os.MkdirTemp("", "host-source")
+		Expect(err).ToNot(HaveOccurred())
+
+		destinationDir, err := os.MkdirTemp("", "host-destination")
+		Expect(err).ToNot(HaveOccurred())
+
+		f, err := os.Create(filepath.Join(sourceDir, file))
+		Expect(err).ToNot(HaveOccurred())
+		err = f.Close()
+		Expect(err).ToNot(HaveOccurred())
+
+		// Get the file stat to check permissions later
+		sourceFileStat, err := os.Stat(filepath.Join(sourceDir, file))
+		Expect(err).ToNot(HaveOccurred())
+
+		err = os.MkdirAll(filepath.Join(sourceDir, directory), 0755)
+		Expect(err).ToNot(HaveOccurred())
+
+		// Get the directory stat to check permissions later
+		sourceDirStat, err := os.Stat(filepath.Join(sourceDir, directory))
+		Expect(err).ToNot(HaveOccurred())
+
+		f, err = os.Create(filepath.Join(sourceDir, directory, fileInDirectory))
+		Expect(err).ToNot(HaveOccurred())
+		err = f.Close()
+		Expect(err).ToNot(HaveOccurred())
+
+		// Get the file in directory stat to check permissions later
+		sourceFileInDirStat, err := os.Stat(filepath.Join(sourceDir, directory, fileInDirectory))
+		Expect(err).ToNot(HaveOccurred())
+
+		By("copy from host to container")
+		name := randomString()
+		i := new(initMachine)
+		session, err := mb.setName(name).setCmd(i.withImage(mb.imagePath).withNow()).run()
+		Expect(err).ToNot(HaveOccurred())
+		Expect(session).To(Exit(0))
+
+		bm := basicMachine{}
+		newImgs, err := mb.setCmd(bm.withPodmanCommand([]string{"pull", TESTIMAGE})).run()
+		Expect(err).ToNot(HaveOccurred())
+		Expect(newImgs).To(Exit(0))
+		Expect(newImgs.outputToStringSlice()).To(HaveLen(1))
+
+		createAlp, err := mb.setCmd(bm.withPodmanCommand([]string{"create", TESTIMAGE, "top"})).run()
+		Expect(err).ToNot(HaveOccurred())
+		Expect(createAlp).To(Exit(0))
+		Expect(createAlp.outputToStringSlice()).To(HaveLen(1))
+
+		containerID := createAlp.outputToStringSlice()[0]
+
+		// Copy a single file into the guest
+		cpFile, err := mb.setCmd(bm.withPodmanCommand([]string{"cp", filepath.Join(sourceDir, file), containerID + ":/tmp/"})).run()
+		Expect(err).ToNot(HaveOccurred())
+		Expect(cpFile).To(Exit(0))
+
+		// Copy a directory into the guest
+		cpDir, err := mb.setCmd(bm.withPodmanCommand([]string{"cp", filepath.Join(sourceDir, directory), containerID + ":/tmp"})).run()
+		Expect(err).ToNot(HaveOccurred())
+		Expect(cpDir).To(Exit(0))
+
+		start, err := mb.setCmd(bm.withPodmanCommand([]string{"start", containerID})).run()
+		Expect(err).ToNot(HaveOccurred())
+		Expect(start).To(Exit(0))
+
+		// Check the single file is created with the appropriate mode, uid, gid
+		exec, err := mb.setCmd(bm.withPodmanCommand([]string{"exec", containerID, "stat", "-c", "%a %u %g", path.Join("/tmp", file)})).run()
+		Expect(err).ToNot(HaveOccurred())
+		Expect(exec).To(Exit(0))
+		execStdOut := exec.outputToStringSlice()
+		Expect(execStdOut).To(HaveLen(1))
+		Expect(execStdOut[0]).To(Equal(fmt.Sprintf("%o %d %d", sourceFileStat.Mode().Perm(), 0, 0)))
```

Comment:
you could simplify this by just using `outputToString()`, I don't see a strong reason to check the line count for all the commands?